components/ws-manager-mk2/service/manager.go

```diff
@@ -271,6 +279,67 @@ func (wsm *WorkspaceManagerServer) StartWorkspace(ctx context.Context, req *wsma
 	}, nil
 }
 
+func isProtectedEnvVar(name string, sysEnvvars []*wsmanapi.EnvironmentVariable) bool {
+	switch name {
+	case "THEIA_SUPERVISOR_TOKENS":
+		return true
+	default:
+		if isGitpodInternalEnvVar(name) {
+			return false
+		}
+		for _, env := range sysEnvvars {
+			if env.Name == name {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+func isGitpodInternalEnvVar(name string) bool {
+	return strings.HasPrefix(name, "GITPOD_") ||
+		strings.HasPrefix(name, "SUPERVISOR_") ||
+		strings.HasPrefix(name, "BOB_") ||
+		strings.HasPrefix(name, "THEIA_") ||
+		name == "NODE_EXTRA_CA_CERTS" ||
+		name == "VSX_REGISTRY_URL"
+}
+
+func (wsm *WorkspaceManagerServer) createWorkspaceSecret(ctx context.Context, owner client.Object, name, namespace string, data map[string]string) error {
+	secret := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		StringData: data,
+	}
+
+	err := wsm.Client.Create(ctx, &secret)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+
+	err = wait.PollWithContext(ctx, 100*time.Millisecond, 5*time.Second, func(c context.Context) (done bool, err error) {
```

Comment:
why the poll actually? If the `Create` succeeded, do we need to check again?

Reply:
Actually not sure why I did that anymore :sweat_smile: . Removed it.